README.md
# OpenIM 1.1


CHANGELOG
########################################
OpenIM 1.1.3
6/2016.0629
添加用户下线行为检测，
	仅仅在用户下线的时候输出日志，用于调试

########################################
OpenIM 1.1.2
5/2016.0628
在mod_receipts_ack中添加用户上线行为检测，
	接收者在上线的时候，可以及时处理cache中的message，
	处理方式是resend
替换原先通过kick功能断开recipient连接的方式，
	改用close
等待“received receipts”的时间缩短为10sec

########################################
OpenIM 1.1.1
4/2016.0623
增强mod_receipts_ack的功能，（效仿XEP-0184的行为）
	cache带有“request receipts”标识的message，
	并在收到“received receipts”标识时，清除相应的cache，
	收到offline行为时，只处理带有“request receipts”标识的message，
		清除相应的cache（如果cache中存在）
	超过3min未收到相应的“received receipts”标识时，
		kick异常recipient，
			jid中含有resource时，只kick相应的recipient，
			jid中未含有resource时，kick全部resource对应的多个recipient，
		将message按offline规则处理，
			offline规则会自动清理cache

########################################
OpenIM 1.0.1
2/2016.0511
针对ejabberd 16.03(4)，p1_xml变更为fast_xml
补充两个本地用户配置工具

########################################
OpenIM 1.0.0
第一版，针对ejabberd 16.01



src/mod_logger.erl
%%%--------------------------------------------------------
%%% File	: mod_logger.erl
%%% Author	: daimaqiao <daimaqiao@126.com>
%%% Purpose	: 实现独立的日志工具
%%% Created	: 2016.3.22
%%% Version	: 2/2016.0511
%%% Dependencies:
%%%		log4erl		独立的日志工具
%%%		mochiweb	log4erl依赖项
%%%	Input	: $EJABBERD_HOME/etc/ejabberd/mod_logger.conf
%%%	Output	: $EJABBERD_HOME/var/log/ejabberd/mod_logger.log
%%%
%%%--------------------------------------------------------

-module(mod_logger).
-author(daimaqiao).
-vsn(2).
-date({2016,5,11}).

-behavior(gen_mod).

-include("logger.hrl").
-include("mod_logger.hrl").
-define(LOGGER, log4erl).

-export([start/2, stop/1]).
-export([mod_opt_type/1]).


start(Host, Opts) ->
	?INFO_MSG("Start module ~p!~n"
			  "Host= ~p~n"
			  "Opts= ~p~n"
			  "Conf= ~p~n"
			  "Log= ~p~n"
			  ,[?MODULE, Host, Opts, get_conf(), get_log()]),
	application:start(?LOGGER),
	?LOGGER:conf(get_conf()),
	?LOGI(" ===== mod_logger is here. ===== "),
	ok.

stop(Host) ->
	?INFO_MSG("Stop module ~p!~n"
			  "Host= ~p~n"
			  ,[?MODULE, Host]),
	ok.


%% get_conf()
%% 返回模块对应的配置文件
get_conf() ->
	Confile= atom_to_list(?MODULE)++".conf",
	get_env_path("EJABBERD_CONFIG_PATH")++Confile.


%% get_log()
%% 返回模块对应的日志文件
get_log() ->
	Logfile= atom_to_list(?MODULE)++".log",
	get_env_path("EJABBERD_LOG_PATH")++Logfile.


%% get_env_path(Key)
%% 返回以“/”结束的目录路径
%% Key: EJABBERD_CONFIG_PATH, EJABBERD_LOG_PATH
%%
get_env_path(Key) ->
	case os:getenv(Key) of
		false-> "./";
		Path ->
			case filename:dirname(Path) of
				"/"-> "/";
				Dir-> Dir++"/"
			end
	end.


%% 不提供模块参数
mod_opt_type(_) ->
	[].



src/xmlel_indent.erl
%%%--------------------------------------------------------
%%% File	: xmlel_indent.erl
%%% Author	: daimaqiao <daimaqiao@126.com>
%%% Purpose	: 为#xmlel补充/清除缩进空格等信息（换行或空格）
%%% Created	: 2016.3.24
%%% Version	: 1/2016.0420
%%% Dependencies:
%%%		fast_xml	原p1_xml，process-one出品的xml工具
%%%		p1_utils	fast_xml依赖项
%%%
%%%--------------------------------------------------------

-module(xmlel_indent).
-author(daimaqiao).
-vsn(1).

-include("jlib.hrl").

-export([inflate_xmlel/1, deflate_xmlel/1]).

-define(INDENT_LEVEL, 4).
-define(MAKE_INDENT(N),{xmlcdata,list_to_binary("\n"++lists:duplicate(N," "))}).
-define(INLINE_MAX, 24).


%% inflate_xml(Xml)
%% 为Xml补充缩进信息
%% 返回扩充之后的#xmlel
%%
inflate_xmlel(Xml) when is_record(Xml, xmlel) ->
	do_inflate_xmlel([], 0, Xml#xmlel.name, Xml#xmlel.attrs, Xml#xmlel.children);
inflate_xmlel(Any) ->
	Any.


%% deflate_xml(Xml)
%% 为Xml清除缩进信息
%% 返回清除缩进之后的#xmlel
%%
deflate_xmlel(Xml) when is_record(Xml, xmlel) ->
	do_deflate_xmlel([], Xml#xmlel.name, Xml#xmlel.attrs, Xml#xmlel.children);
deflate_xmlel(Any) ->
	Any.


%% do_deflate_xmlel(Acc, Name, Attrs, Children)
%% 在Children中清除缩进信息（去除非可见文字信息）
%% Acc是累加器
%% 返回不含缩进信息的#xmlel{name=Name, attrs=Attrs, children=NewChildren}
%%
do_deflate_xmlel([], Name, Attrs, []) ->
	#xmlel{name= Name, attrs= Attrs, children= []};
do_deflate_xmlel(Acc, Name, Attrs, []) ->
	Children= lists:reverse(Acc),
	#xmlel{name= Name, attrs= Attrs, children= Children};
do_deflate_xmlel(Acc, Name, Attrs, [{xmlcdata, B}| Children]) ->
	case re:replace(B, "(^\\s+)|(\\s+$)", "", [global,{return,binary}]) of
		<<"">> ->
			do_deflate_xmlel(Acc, Name, Attrs, Children);
		X ->
			Child= {xmlcdata, X},
			do_deflate_xmlel([Child| Acc], Name, Attrs, Children)
	end;
do_deflate_xmlel(Acc, Name, Attrs, [Child| Children]) ->
	NewChild= deflate_xmlel(Child),
	do_deflate_xmlel([NewChild| Acc], Name, Attrs, Children).


%% do_inflate_xmlel(Acc, N, Name, Attrs, Children)
%% 在Children中补充缩进信息
%% Acc是累加器，N表示缩进量（空格数量）
%% 返回包含缩进信息的#xmlel{name=Name, attrs=Attrs, children=NewChildren}
%%
do_inflate_xmlel([], _, Name, Attrs, []) ->
	#xmlel{name= Name, attrs= Attrs, children= []};
do_inflate_xmlel([], _, Name, Attrs, Children= [{xmlcdata, B}])
  when size(B)=< ?INLINE_MAX ->
	#xmlel{name= Name, attrs= Attrs, children= Children};
do_inflate_xmlel(Acc, N, Name, Attrs, []) ->
	Children= lists:reverse([?MAKE_INDENT(N)| Acc]),
	#xmlel{name= Name, attrs= Attrs, children= Children};
do_inflate_xmlel(Acc, N, Name, Attrs, [Child | Children]) ->
	Indent= N+ ?INDENT_LEVEL,
	NewChild= do_inflate_xmlel2(Indent, Child),
	Acc1= [?MAKE_INDENT(Indent)| Acc],
	NewAcc= [NewChild| Acc1],
	do_inflate_xmlel(NewAcc, N, Name, Attrs, Children).

%% do_inflate_xmlel2(N, X)
%% 根据X的类型决定是否为X补充缩进信息
%% 返回更新后的X
%%
do_inflate_xmlel2(N, X) when is_record(X, xmlel) ->
	do_inflate_xmlel([], N, X#xmlel.name, X#xmlel.attrs, X#xmlel.children);
do_inflate_xmlel2(_, X) ->
	X.



src/mod_dump_packet.erl
%%%--------------------------------------------------------
%%% File	: mod_dump_packet.erl
%%% Author	: daimaqiao <daimaqiao@126.com>
%%% Purpose	: 输出XMPP包内容到文件
%%% Created	: 2016.3.22
%%% Version	: 7/2016.0617
%%% Dependencies:
%%%		mod_logger			log4erl的模块封装
%%%		log4erl				独立的日志工具
%%%		mochiweb			log4erl依赖项
%%%		fast_xml			原p1_xml，xml工具
%%%		p1_utils			fast_xml依赖项
%%%		xmlel_indent.erl	为#xmlel添加缩进空格等信息
%%%
%%%--------------------------------------------------------

-module(mod_dump_packet).
-author(daimaqiao).
-vsn(7).
-date({2016,6,17}).

-behavior(gen_mod).

-include("logger.hrl").
-include("jlib.hrl").
-include("mod_logger.hrl").

-define(OPT_STANZA, dump_packet_stanza).

-export([start/2, stop/1]).
-export([install_dump_packet/0, dump_packet/1]).
-export([mod_opt_type/1]).

start(Host, Opts) ->
	?INFO_MSG("Start module ~p!~n"
			  "Host= ~p~n"
			  "Opts= ~p~n"
			  ,[?MODULE, Host, Opts]),
	cache_opts(Opts),
	install_dump_packet(),
	ok.

stop(Host) ->
	?INFO_MSG("Stop module ~p!~n"
			  "Host= ~p~n"
			  ,[?MODULE, Host]),
	ok.


%% install_dump_packet()
%% 设置filter_packet事件
%%
install_dump_packet() ->
    ejabberd_hooks:add(filter_packet, global, ?MODULE, dump_packet, 20).


%% dump_packet(From, To, Packet)
%% 返回drop表示丢弃，否则就原样返回
%%
dump_packet(All= {From, To, Packet}) ->
	case select_packet(From#jid.user, To#jid.user, Packet#xmlel.name) of
		true ->
			dump_to_text(All);
		_ ->
			ok
	end,
	All.

%% select_packet(User1, User2, Stanza)
%% 选择匹配的数据包，返回true表示选中
%%
select_packet(User1, User2, Stanza) ->
	{OptUser, OptStanza}= fetch_opts_user_stanza(),

	((OptUser== []) or lists:member(User1, OptUser) or lists:member(User2, OptUser)) and
	((OptStanza== []) or lists:member(Stanza, OptStanza)).


%% do_dump_packet(From, To, Packet)
%% 原样输出包内容
%%
do_dump_packet({From, To, Packet}) ->
	?LOGD(" *** dumping ***~n"
		  "  From  = ~ts~n"
		  "  To    = ~ts~n"
		  "  Packet= ~ts~n"
		  "~n"
		  ,[From, To, Packet]),
	ok.

%% dump_to_text(From#jid{}, To#jid{}, Packet#xmlel{})
%% 将内容转换成字符串再输出
%%
dump_to_text({From= #jid{}, To= #jid{}, Packet= #xmlel{}}) ->
	Id1= jlib:jid_to_string(From),
	Id2= jlib:jid_to_string(To),
	Xmlel1= xmlel_indent:deflate_xmlel(Packet),
	Xmlel2= xmlel_indent:inflate_xmlel(Xmlel1),
	Xml= fxml:element_to_binary(Xmlel2),
	do_dump_packet({binary_to_list(Id1), binary_to_list(Id2),
					binary_to_list(Xml)});
dump_to_text(All) ->
	do_dump_packet(All).


%%
%% mod_opt_type
%% 识别并转换模块参数
%% user:   [string(), ...]
%% stanza: [string(), ...]
%%
mod_opt_type(user) ->
	fun(A) -> A end;
mod_opt_type(stanza) ->
	fun(A) -> A end;
mod_opt_type(_) ->
    [user, stanza].


%% cache_opts(Opts)
%% 将必要参数保存到ETS中
%% 多个hosts共享一份
%% 返回ok
%%
cache_opts(Opts) ->
	case lists:member(?MODULE, ets:all()) of
		true ->
			%% 参数只缓存一次
			ok;
		_ ->
			ets:new(?MODULE, [set, protected, named_table]),
			User= opt_of(user, Opts, []),
			Stanza= opt_of(stanza, Opts, []),
			ets:insert(?MODULE, {opts_user_stanza, {User, Stanza}}),
			ok
	end.

%% opt_of(Key, Opts, Def)
%% 从列表Opts中找到Key，返回Key对应的Val，或者Def
%%
opt_of(Key, Opts, Def) ->
	case lists:keyfind(Key, 1, Opts) of
		{Key, Val} -> Val;
		false -> Def
	end.

%% fetch_opts_user_stanza()
%% 从ETS中取回参数
%% 返回{User::list(), Stanza::list()}
%%
fetch_opts_user_stanza() ->
	case ets:lookup(?MODULE, opts_user_stanza) of
		[{opts_user_stanza, {User, Stanza}}] ->
			{User, Stanza};
		_ ->
			{[], []}
	end.



src/mod_receipts_ack.erl
%%%--------------------------------------------------------
%%% File	: mod_receipts_ack.erl
%%% Author	: daimaqiao <daimaqiao@126.com>
%%% Purpose	: 实现XEP-0184协议的服务端应答
%%% Created	: 2016.3.29
%%% Version	: 8/2016.0629
%%% Dependencies:
%%%		log4erl		独立的日志工具
%%%		mochiweb	log4erl依赖项
%%%		fast_xml	原p1_xml，xml工具
%%%		p1_utils	fast_xml依赖项
%%%	Input	: $EJABBERD_HOME/etc/ejabberd/mod_logger.conf
%%%	Output	: $EJABBERD_HOME/var/log/ejabberd/mod_logger.log
%%%
%%%	References:
%%%		https://github.com/Mingism/ejabberd-stanza-ack
%%%		https://github.com/mrDoctorWho/ejabberd_mod_apns
%%%
%%%--------------------------------------------------------

-module(mod_receipts_ack).
-author(daimaqiao).
-vsn(8).
-date({2016,6,29}).

-behavior(gen_mod).

-include("logger.hrl").
-include("jlib.hrl").
-include("mod_logger.hrl").

-define(NS_RECEIPTS, <<"urn:xmpp:receipts">>).
-define(ACK_PREFIX, <<"ack.">>).

%%-define(MODULE_RESEND, list_to_atom(atom_to_list(?MODULE)++"resend")).
-define(MODULE_RESEND, mod_receipts_ack_resend).


-define(TIMER_GO_OFFLINE, (10*1000)).%% 10s
-define(POS_KEY, 1).
-define(POS_STATE, 2).
-define(POS_TIMER, 3).
-define(POS_TS, 4).
-define(POS_ORIG, 5).


-export([start/2, stop/1]).
-export([install_receipts_ack/1, receipts_ack/4, offline_ack/3, logon_ack/3, logoff_ack/3]).
-export([mod_opt_type/1]).
-export([go_offline/4]).


start(Host, Opts) ->
	?INFO_MSG("Start module ~p!~n"
			  "Host= ~p~n"
			  "Opts= ~p~n"
			  ,[?MODULE, Host, Opts]),
	cache_init(),
	install_receipts_ack(Host),
	ok.

stop(Host) ->
	?INFO_MSG("Stop module ~p!~n"
			  "Host= ~p~n"
			  ,[?MODULE, Host]),
	uninstall_receipts_ack(Host),
	ok.


%% install_receipts_ack()
%% 设置filter_packet事件
%%
install_receipts_ack(Host) ->
	mod_disco:register_feature(Host, ?NS_RECEIPTS),
	ejabberd_hooks:add(offline_message_hook, Host, ?MODULE, offline_ack, 20),
	ejabberd_hooks:add(user_send_packet, Host, ?MODULE, receipts_ack, 20),
	ejabberd_hooks:add(sm_register_connection_hook, Host, ?MODULE, logon_ack, 20),
	ejabberd_hooks:add(sm_remove_connection_hook, Host, ?MODULE, logoff_ack, 20).
%%
uninstall_receipts_ack(Host) ->
	mod_disco:unregister_feature(Host, ?NS_RECEIPTS),
	ejabberd_hooks:delete(offline_message_hook, Host, ?MODULE, offline_ack, 20),
	ejabberd_hooks:delete(user_send_packet, Host, ?MODULE, receipts_ack, 20),
	ejabberd_hooks:delete(sm_register_connection_hook, Host, ?MODULE, logon_ack, 20),
	ejabberd_hooks:delete(sm_remove_connection_hook, Host, ?MODULE, logoff_ack, 20).


logoff_ack(_Sid, Jid, _Info) ->
	KeyResend= cache_key(Jid),
	Found= ets:lookup(?MODULE_RESEND, KeyResend),
	?LOGD("just show receipts are waiting for on logoff: key= ~p, found= ~p", [jid:to_string(Jid), Found]).

logon_ack(_Sid, Jid, _Info) ->
	KeyResend= cache_key(Jid),
	Found= ets:take(?MODULE_RESEND, KeyResend),
	?LOGD("check receipts are waiting for on logon: key= ~p, found= ~p", [jid:to_string(Jid), Found]),
    lists:foreach(
		fun({_, Key}) ->
			cancel_timer_and_resend(Key)
		end, Found),
	ok.


offline_ack(From, To, Packet= #xmlel{name= <<"message">>, attrs= Attrs}) ->
	case fxml:get_attr_s(<<"type">>, Attrs) of
		<<"chat">> ->
			offline_ack_request(From, To, Packet, Attrs);
		_ ->
			%% ignore non-chat-type
			false
	end,
	ok;
offline_ack(_,_,_) ->
	ok.

%% 处理完成返回ok，不匹配未处理返回false
offline_ack_request(From, To, Message, Attrs) ->
	case fxml:get_subtag_with_xmlns(Message, <<"request">>, ?NS_RECEIPTS) of
		false ->
			%% mismatching
			false;
		_ ->
			Id= fxml:get_attr_s(<<"id">>, Attrs),
			?LOGD("get offline message, and receipts is required: key= ~p", [cache_key(From, Id)]),
			cache_remove_chat(From, To, Id),
			do_receipts_ack(To, From, Id),
			ok
	end.



%% receipts_ack(#xmlel(), C2SStage?, #jid(), #jid()) ->
%% #xmlel()
receipts_ack(Packet= #xmlel{name= <<"message">>, attrs= Attrs}, _C2SStage, From, To) ->
	case fxml:get_attr_s(<<"type">>, Attrs) of
		<<"chat">> ->
			case receipts_ack_request(From, To, Packet, Attrs) of
				false ->
					case receipts_ack_received(From, To, Packet) of
						ok ->
							ok;
						_ ->
							false
					end;%% received
				_ ->
					ok
			end;%% request
		_ ->
			%% ignore non-chat-type
			false
	end,
	Packet;
receipts_ack(Packet, _, _, _) ->
	Packet.

%% receipts_ack_request(#jid(), #jid(), #xmlel(), list()) ->
%% ok | false
%% 处理完成返回ok，不匹配未处理返回false
receipts_ack_request(From, To, Message, Attrs) ->
	case fxml:get_subtag_with_xmlns(Message, <<"request">>, ?NS_RECEIPTS) of
		false ->
			%% mismatching
			false;
		_ ->
			Id= fxml:get_attr_s(<<"id">>, Attrs),
			%%Body= fxml:get_subtag_cdata(Message, <<"body">>),
			?LOGD("get chat message, and receipts is required: key= ~p", [cache_key(From, Id)]),
			cache_save_chat(From, To, Message, Id)
	end.

%% receipts_ack_received(#jid(), #jid(), #xmlel()) ->
%% ok | false
%% 处理完成返回ok，不匹配未处理返回false
receipts_ack_received(From, To, Message) ->
  	case fxml:get_subtag_with_xmlns(Message, <<"received">>, ?NS_RECEIPTS) of
  		#xmlel{attrs= ReceivedAttrs} ->
  			case fxml:get_attr(<<"id">>, ReceivedAttrs) of
  				{value, Id} ->
					?LOGD("receipts is received: key= ~p", [cache_key(To, Id)]),
					cache_remove_chat(To, From, Id),
  					ok;
  				_ ->
					false
  			end;
  		_ ->
			false
  	end.


%% cache_key(#jid(), binary()) ->
%% {binary(), binary()}
cache_key(Jid, Id) ->
	{jid:to_string(Jid), Id}.
%% cache_key(#jid()) ->
%% {binary()}
cache_key(Jid) ->
	jid:to_string(jid:remove_resource(Jid)).



cache_init() ->
	case lists:member(?MODULE, ets:all()) of
		true ->
			%% 初始化缓存一次
			ok;
		_ ->
			timer_init(),
			ets:new(?MODULE, [set, public, named_table]),
			ets:new(?MODULE_RESEND, [bag, public, named_table]),
			ok
	end.

%% cache_save_chat
%%	(#jid(), #jid(), #xmlel(), binary()) ->
%% true
cache_save_chat(From, To, Packet, Id) ->
    Resources= ejabberd_sm:get_user_resources(To#jid.user, To#jid.server),
	?LOGD("there are ~p recipients of ~p online: resources= ~p",
		  [length(Resources), jid:to_string(To), Resources]),
	Ts= erlang:system_time(seconds),
	Key= cache_key(From, Id),
	KeyResend= cache_key(To),
	Timer= case timer_reset(Key, ?TIMER_GO_OFFLINE, go_offline, [Key, From, To, Packet]) of
			   {ok, T} ->
					?LOGD("chat message is being cached, waiting for receipts: timer= ~p, key= ~p", [T, Key]),
				   T;
			   {error, Reason} ->
				   ?LOGW("FAILED to set timer, when caching chat message: ~p", [Reason]),
				   false
		   end,
	ets:insert(?MODULE, {Key, {state, save_chat}, {timer, Timer}, {ts, Ts}, {orig, {From, To, Packet}}}),
	ets:insert(?MODULE_RESEND, {KeyResend, Key}).

%% cache_remove_chat(#jid(), #jid(), binary())->
%% true
cache_remove_chat(From, To, Id) ->
	Key= cache_key(From, Id),
	KeyResend= cache_key(To),
	timer_unset(Key),
	ets:delete(?MODULE, Key),
	ets:delete_object(?MODULE_RESEND, {KeyResend, Key}).


%% go_offline(tuple(), #jid(), #jid(), #xmlel()) ->
%% ok | false
%% clear timer, save into offline, and drop bad recipient
go_offline(Key, From, To, Packet) ->
	%% clear timer
	?LOGD("time up! the recipient may be not online, because of none receipts: key= ~p", [Key]),
	Ts= erlang:system_time(seconds),
	try ets:update_element(?MODULE, Key, [{?POS_STATE, {state, go_offline}},
										  {?POS_TIMER, {timer, timeout}},
										  {?POS_TS, {ts, Ts}}])
	catch
		_:Reason ->
			%% may not exist
			?LOGW("FAILED to clear the timer waiting for receipts after timeout: key= ~p~nerror= ~p", [Key, Reason]),
			false
	end,
	%% save into offline
	save_into_offline(From, To, Packet),
	%% drop bad recipient
	drop_recipient(To).


timer_init() ->
	timer:start().

%% timer_unset(tuple()) ->
%% ok | false
%% clear last timer if exists
timer_unset(Key) ->
	try ets:lookup_element(?MODULE, Key, ?POS_TIMER) of
	{timer, Timer} when is_tuple(Timer) ->
		%% cancel the timer
		?LOGD("unset timer ~p for ~p", [Timer, Key]),
		timer:cancel(Timer),
		ok;
	_ ->
		%% nothing
		false
	catch
		_:_ ->
			%% may not exist
			false
	end.

%% timer_reset(tuple(), integer(), atom(), list()) ->
%% {ok, TRef} | {error, Reason}
%% clear last timer if exists, and then create a new timer
timer_reset(Key, Timeout, Action, Args) ->
	timer_unset(Key),
	timer:apply_after(Timeout, ?MODULE, Action, Args).


%% cancel_timer_and_resend(tuple()) ->
%% ok | false
cancel_timer_and_resend(Key) ->
	case ets:lookup(?MODULE, Key) of
		[{Key, _, {timer, Timer}, _, {orig, {From, To, Packet}}}] when is_tuple(Timer) ->
			%% cancel the timer
			?LOGD("cancel timer ~p for ~p", [Timer, Key]),
			timer:cancel(Timer),
			%% update ets
			Ts= erlang:system_time(seconds),
			try ets:update_element(?MODULE, Key, [{?POS_STATE, {state, to_resend}},
												  {?POS_TIMER, {timer, cancel}},
												  {?POS_TS, {ts, Ts}}])
			catch
				_:Reason ->
					%% may not exist
					?LOGW("FAILED to clear the timer that is canceled: key= ~p~nerror= ~p", [Key, Reason]),
					false
			end,
			%% resend the message
			?LOGD("resend message in 5sec: ~p --> ~p", [jid:to_string(From), jid:to_string(To)]),
			timer:apply_after(5000, ejabberd_router, route, [From, To, Packet]);
%%			ejabberd_router:route(From, To, Packet);
		Found ->
			?LOGD("timer waiting for receipts may be already cleared, resending is not required!~nfound= ~p", [Found]),
			false
	end.



%% (#jid())->
%% ok | false
drop_recipient(Jid=#jid{luser=User,lserver=Server,lresource= <<"">>}) ->
    Resources= ejabberd_sm:get_user_resources(User, Server),
    lists:foreach(
		fun(Resource) ->
			case ejabberd_sm:get_session_pid(User, Server, Resource) of
				none ->
					false;
				Pid ->
					close_pid(Pid)
			end
		end, Resources),
	if 
		length(Resources) > 0 ->
			?LOGI("DROP multiple recipients for nonresponse: jid= ~p, resources= ~p",
				  [jid:to_string(Jid), Resources]),
			ok;
		true ->
			false
	end;
drop_recipient(Jid=#jid{luser=User,lserver=Server,lresource=Resource}) ->
	case ejabberd_sm:get_session_pid(User, Server, Resource) of
		none ->
			false;
		Pid ->
			?LOGI("DROP recipient for nonresponse: jid= ~p", [jid:to_string(Jid)]),
			close_pid(Pid)
	end.

close_pid(Pid) ->
	ejabberd_c2s:close(Pid),
	ok.


%% (#jid(), #jid(), #xmlel()) ->
%% ok
save_into_offline(From, To, Packet) ->
	?LOGD("message is going into the offline storage: ~p --> ~p",
		  [jid:to_string(From), jid:to_string(To)]),
	ejabberd_hooks:run(offline_message_hook, From#jid.server, [From, To, Packet]).


%% 仅当消息进入离线队列，无法收到接收方receipts时
%% 服务器回应ack receipts
%% 参照XEP-0184协议，向To回应服务器端代From发的Ack
%% 返回ok
do_receipts_ack(#jid{luser=User,lserver=Server,lresource=Resource}, To, Id) ->
	Jid= jid:make(User, <<?ACK_PREFIX/binary, Server/binary>>, Resource),
	?LOGD("make ack of receipts locally: ~p --> ~p, id= ~p",
		  [jid:to_string(Jid), jid:to_string(To), Id]),
	Xmlel= #xmlel{name= <<"message">>, attrs= [{<<"to">>, To}],
				  children= [#xmlel{name= <<"received">>,
									attrs= [{<<"xmlns">>, ?NS_RECEIPTS},
											{<<"id">>, Id}],
									children= []}]},
    ejabberd_router:route(Jid, To, Xmlel).


%% 不提供模块参数
mod_opt_type(_) ->
	[].

